### PR TITLE
persist: turn Persist::persist() operator into modular persist()/seal() operators

### PR DESCRIPTION
**Pull request by aljoscha**
### Motivation

This turns the monolithic persist() operator, which was responsible for
both persisting and replaying persisted data into a simpler persist()
operator that only persists. We also add a seal() operator that can seal
a collection based on the advancing frontier.

These are also more efficient (less blocking) because we don't recv(),
that is block, on every write. Instead, we now buffer write futures and
wait on them when the frontier advances.

We will use these operators in the implementation of
persistent_upsert(), and when implementing timestamp binding persistence
for sources.

### Description

The commit message and rustdoc hopefully explain the changes. Please let me know if they don't!

Check my wip branch for context: https://github.com/aljoscha/materialize/commits/wip-persistent-kafka-source. Especially the commits around `persistent_upsert()`.

### Checklist

- [x] This PR has adequate test coverage.
- [x] This PR adds a release note for any user-facing behavior changes.

### CODE REVIEW

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:56]: frank has mentioned a couple times that it's also appropriate for us to emit a single stream of Result instead of an ok and err stream. fyi in case that ends up making anything easier for you

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:77]: nit: the persist crate tends to use capitalization and punctuation in comments. here and elsewhere

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:103]: should we use a BTreeMap instead of a HashMap? maybe not given my above comment

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:102]: I'm uncomfortable hard coding single element antichains where it's avoidable. I think this just wants to iterate the keys and keep them if `!input_frontier.less_equal(key)` no?

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:191]: comment here plz

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:194]: I actually don't follow what's happening here, probably because I haven't directly used OperatorBuilder. perhaps some context?

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:218]: ditto hardcoding single element antichains

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:240]: woah is this really how allow compaction works in sources?

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs:103]: Which one was the comment "above"? On github that's not always clear. :sweat_smile:

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs:240]: It is not! In the real code the coordinator tells workers when they can compact. See here: https://github.com/MaterializeInc/materialize/blob/41b566c1704cbf5b46f1636b98a9065e3c10e513/src/dataflow/src/server.rs#L932. We have to do something similar, where workers have a mapping to write handles for which they need to allow compaction.

I had this code in there to see actual compaction going on in the mock Kafka example. I'll remove from this code and maybe add compaction as an explicit thing in the example.

Thinking more broadly, I think with persistence we wouldn't even need the coordinator to tell the workers that they can compact. The coordinator could talk to persistence directly. Follow-up from that: if/when the coordinator, workers, ingestion, and persistence are separate (possibly distributed) processes, can there be multiple writers to a persistent collection? Right now locking would prevent that, I think?

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs:194]: We need to maintain a capability that allows us to send errors downstream. We can either do that by keeping and managing the initial capability or by retaining capabilities from input data.

I'll push a change that retains input capabilities, because that seems to be more idiomatic. I don't know if it simplifies things, though... :sweat_smile:

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs:218]: Agreed! Should I change this code to iterate through all elements and seal? Or should we change `seal()` to take an `Antichain<u64>` instead of a single `u64`. I know that internally, `seal()` does a `Antichain::from_elem(ts)` right now.

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs:194]: It's probably best to to a quick huddle (in slack or elsewhere) about this.

Comment (danhhz) [reply src/persist/src/operators/stream.rs:194]: Oh I think you're overestimating the sophistication of my question :). After seeing your question in dataflow slack, I think the answer I was looking for here is "these are capabilities that correspond to the operator inputs".

Like frank said there, I think the right thing to do is keep whichever capabilities make the most semantic sense, in this case the operator is not really a source, so what you have here is correct I think

File: src/persist/src/operators/stream.rs
Comment (danhhz) [reply src/persist/src/operators/stream.rs:218]: I'm fine with making seal take an Antichain<u64>. Happy to pick that off and let you rebase on top of it if you don't want to deal with it

File: src/persist/src/operators/stream.rs
Comment (danhhz) [reply src/persist/src/operators/stream.rs:240]: > I'll remove from this code and maybe add compaction as an explicit thing in the example.

Yeah this sounds right to me

> Thinking more broadly, I think with persistence we wouldn't even need the coordinator to tell the workers that they can compact. The coordinator could talk to persistence directly.

This is one of the things I'm hoping gets figured out in the architecture/correctness discussions that frank is pushing on 🤞 

> Follow-up from that: if/when the coordinator, workers, ingestion, and persistence are separate (possibly distributed) processes, can there be multiple writers to a persistent collection? Right now locking would prevent that, I think?

Writes and seals, I think no. If there's a reason you think we need that, let's definitely jump on this sooner than later because that's a huge violation of assumptions I'm making.

allow_compaction is possibly a different story: IIUC the way this works in arranged/trace is that the compaction frontier is only advanced to some point once all readers have advanced it to that point. It's easy to imagine that we'd want our analog to match. I'm sort of hoping to punt on this for a while while being prepared to revisit it if in a services world if necessary. 

One complication here is the idea of arjun's chain of materialize instances (also applicable to blue/green deployments etc). A chained mz instance using a persistent collection as a source IMO should not be able to _in any way_ affect the performance of the cluster that's writing the collection. That includes holding up compactions

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs]: Sure, I guess it will be quicker when you do it. :ok_hand:

Comment (aljoscha) [reply src/persist/src/operators/stream.rs]: One thing to note, though. There is an asymmetry in timely itself, where input frontiers are antichains, but capabilities are usually just for a timestamp. So an operator would, say, have `Antichain<u64>` as input frontiers but hold on to `Capability<u64>`, where `capability.downgrade(&u64)` roughly corresponds to `seal(&u64)`.

Comment (danhhz) [reply src/persist/src/operators/stream.rs]: Okay that's interesting. The closest analog to `seal` that exists in timely/differential is the way dd's Arrange operator moves things from its internal buffer into a Trace, so lets see how that works...

It looks like that happens by adding a Batch to the trace where the Batch's lower is required to match the current upper of the trace. In practice, I think this batch's upper antichain only ever differs from its lower by a single element, but it looks like this is a bit of awkwardness due to exactly the timely asymmetry you mention:

- https://docs.rs/differential-dataflow/0.12.0/src/differential_dataflow/operators/arrange/arrangement.rs.html#586-589
- https://docs.rs/differential-dataflow/0.12.0/src/differential_dataflow/operators/arrange/arrangement.rs.html#602-605

I haven't thought through all the implications of this, but given that you're going to have to advance the frontier/drop capabilities after the seal succeeds, maybe it is more natural to iterate over the elements and individually seal the ones that have advanced?

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs]: Oh, but these capabilities are capabilities for the outputs. So I was doing the thing that Frank recommended we don't do for non-source operators. (to many double-no's in there...)

I pushed a change that retains input capabilities, but the problem with that is that we can only seal once we have received data at least once. Before that, we didn't yet have a chance to retain a capability.

Side note: the code I use now for handling the capabilities comes mostly from the dd arrange operator. It's somewhat complicated but very general.

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs]: I think so as well. I pushed a commit that does that.

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs]: something something clone_from? (this is a shot in the dark, I'm not sure if it's relevant here)

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:233]: Could this instead subsume the `if progress` check?

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:88]: nit: is it possible to get this `1` constant from somewhere? taking a wild stab maybe `data_output_stream.name.port`? it's fine if the answer is no

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:100]: nit: can we invert the control flow here to make this `if progress { return }`? dunno about others, but I personally find that style easier to reason about

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:222]: same nit: can we invert the control flow here to be `if !active_seal_operator { return }`?

File: src/persist/src/operators/stream.rs
Comment (danhhz) [inline src/persist/src/operators/stream.rs:228]: and again :D

**Review by danhhz — APPROVED** [review]
LGTM! This is great!

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs]: It's a good shot in the dark, but the types don't line up. And this idiom is also used elsewhere in the code, when managing input frontiers.

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs:88]: Excellent!

File: src/persist/src/operators/stream.rs
Comment (aljoscha) [reply src/persist/src/operators/stream.rs:233]: I'm also managing the capabilities and updating the input frontier only if there is progress. So I think we still need that in addition to checking each frontier element.

**Comment by aljoscha** [comment]
Thanks a lot for the reviews! I'm inverting the control flows before merging and also got rid of the `1` constants.

**Pull request merged**
